### ISSUE TITLE
SF_SWITCH_FALLTHROUGH False positives

### ISSUE THREAD

**Issue by zbynek** [spotbugs/spotbugs#3767]
With update from 4.9.2 to 4.9.6 there were several reports of  false positives for this rule and the related `SF_DEAD_STORE_DUE_TO_SWITCH_FALLTHROUGH`. 

In the code below I picked 3 examples. I'm not sure to what extent are they related, so I'm putting them all in this one issue.

```java
package ghIssues;

import java.io.IOException;

public class Issue3767 {

    double m;
    public void check(int state) {

        switch (state) {
            default:
                return;
            case 1:
                m = m + 1;
                break;
            case 2:
                break;
        }
    }

    public void printNumbers(int size, int type) {
        switch (type) {
            case 1:
                short k = 1;
                while (k < size - 1) {
                    System.out.println(k);
                    k++;
                }
                break;
            case 2:
                k = 1;
                while (k < size - 1) {
                    System.out.println(k + 1);
                    k++;
                }
                break;
            default:
                // should not happen
                break;
        }
    }

    final public String term(int state) throws IOException {

        switch (state) {
            case 1:{
                label_5: while (true) {
                    switch (2 * state) {
                        case 8:{
                            ;
                            break;
                        }
                        default:
                            m = 2;
                            break label_5;
                    }
                    m = 1;
                }
                break;
            }
            case 2:{
                label_6:
                while (true) {
                    switch (2 * state) {
                        case 8:{
                            ;
                            break;
                        }
                        default:
                            m = 2;
                            break label_6;
                    }
                    m = 1;
                }
                break;
            }
            default:
                m = 7;
                throw new IOException();
        }
        return "";
    }

}

```

For the first one the workaround is to put the `default` branch last, for the second one it's possible to extract the switch branches to their own methods, for the third one that would probably also help, but it's part of generated parser code where the nested switches cannot be avoided.
